// @squarecloud/api/lib/services/index.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __publicField = (obj, key, value) => __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);

// src/services/index.ts
var services_exports = {};
__export(services_exports, {
  APIService: () => APIService,
  ApplicationCacheService: () => ApplicationCacheService,
  GlobalCacheService: () => GlobalCacheService
});
module.exports = __toCommonJS(services_exports);

// src/structures/application/base.ts
var import_form_data2 = __toESM(require("form-data"));

// src/assertions/literal.ts
var z = __toESM(require("zod"));
var StringSchema = z.coerce.string();
var BooleanSchema = z.coerce.boolean();
var PathLikeSchema = z.string().or(z.instanceof(Buffer));

// src/modules/applications.ts
var import_form_data = __toESM(require("form-data"));

// src/services/cache/base.ts
var BaseCacheService = class {
  constructor() {
    __publicField(this, "cache");
  }
  set(key, value) {
    Reflect.set(this, key, value);
  }
  get(key) {
    return this.cache[key];
  }
  remove(key) {
    Reflect.set(this.cache, key, void 0);
  }
};

// src/services/cache/application.ts
var ApplicationCacheService = class extends BaseCacheService {
  constructor() {
    super(...arguments);
    __publicField(this, "cache", {
      status: void 0,
      backups: void 0,
      logs: void 0
    });
  }
  get status() {
    return this.cache.status;
  }
  get backups() {
    return this.cache.backups;
  }
  get logs() {
    return this.cache.logs;
  }
};

// src/structures/error.ts
var SquareCloudAPIError = class extends TypeError {
  constructor(code, message, options) {
    super(code);
    this.name = "SquareCloudAPIError";
    this.message = (code?.replaceAll("_", " ").toLowerCase().replace(/(^|\s)\S/g, (L) => L.toUpperCase()) || "UNKNOWN_CODE") + (message ? `: ${message}` : "");
    if (options?.stack) {
      this.stack = options.stack;
    }
    if (options?.cause) {
      this.cause = options.cause;
    }
  }
};

// src/services/api.ts
var APIService = class {
  constructor(apiKey) {
    this.apiKey = apiKey;
    __publicField(this, "baseUrl", "https://api.squarecloud.app");
    __publicField(this, "version", "v2");
    __publicField(this, "userId");
    this.userId = apiKey.split("-")[0];
  }
  async request(...[path, options]) {
    const { url, init } = this.parseRequestOptions(path, options);
    const response = await fetch(url, init).catch((err) => {
      throw new SquareCloudAPIError(err.code, err.message);
    });
    if (response.status === 413) {
      throw new SquareCloudAPIError("PAYLOAD_TOO_LARGE", "Payload too large");
    }
    if (response.status === 429) {
      throw new SquareCloudAPIError(
        "RATE_LIMIT_EXCEEDED",
        "Rate limit exceeded"
      );
    }
    if (response.status === 502 || response.status === 504) {
      throw new SquareCloudAPIError("SERVER_UNAVAILABLE", "Server unavailable");
    }
    const data = await response.json().catch(() => {
      throw new SquareCloudAPIError("CANNOT_PARSE_RESPONSE", "Try again later");
    });
    if (!data || data.status === "error" || !response.ok) {
      throw new SquareCloudAPIError(data?.code || "COMMON_ERROR");
    }
    return data;
  }
  parseRequestOptions(path, options) {
    const init = options || {};
    init.method = init.method || "GET";
    init.headers = {
      Accept: "application/json",
      ...init.headers || {},
      Authorization: this.apiKey
    };
    const url = new URL(path, `${this.baseUrl}/${this.version}`);
    if ("query" in init && init.query) {
      const query = new URLSearchParams(init.query);
      url.search = query.toString();
      init.query = void 0;
    }
    if ("body" in init && init.body && !(init.body instanceof Buffer)) {
      init.body = JSON.stringify(init.body);
      init.headers = {
        ...init.headers,
        "Content-Type": "application/json"
      };
    }
    return { url, init };
  }
};

// src/services/cache/global.ts
var GlobalCacheService = class extends BaseCacheService {
  constructor() {
    super(...arguments);
    __publicField(this, "cache", {
      user: void 0
    });
  }
  get user() {
    return this.cache.user;
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  APIService,
  ApplicationCacheService,
  GlobalCacheService
});
