// @squarecloud/api-types/payloads/v2/application.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// payloads/v2/application.ts
var application_exports = {};
__export(application_exports, {
  ApplicationLanguage: () => ApplicationLanguage
});
module.exports = __toCommonJS(application_exports);
var ApplicationLanguage = {
  JavaScript: "javascript",
  TypeScript: "typescript",
  Python: "python",
  Java: "java",
  Elixir: "elixir",
  Go: "go",
  Rust: "rust",
  PHP: "php",
  Dotnet: "dotnet",
  Static: "static"
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ApplicationLanguage
});
