// @squarecloud/api-types/v2.mjs
// common/v2.ts
var UserId = (id) => id;
var ApplicationId = (id) => id;

// payloads/v2/application.ts
var ApplicationLanguage = {
  JavaScript: "javascript",
  TypeScript: "typescript",
  Python: "python",
  Java: "java",
  Elixir: "elixir",
  Go: "go",
  Rust: "rust",
  PHP: "php",
  Dotnet: "dotnet",
  Static: "static"
};

// payloads/v2/deploy.ts
var DeploymentState = {
  Pending: "pending",
  Clone: "clone",
  Success: "success",
  Error: "error"
};

// payloads/v2/files.ts
var FileType = {
  File: "file",
  Directory: "directory"
};

// payloads/v2/network.ts
var APINetworkDNSStatus = {
  Pending: "pending",
  PendingValidation: "pending_validation",
  Active: "active"
};

// payloads/v2/status.ts
var ApplicationStatus = {
  Exited: "exited",
  Created: "created",
  Starting: "starting",
  Restarting: "restarting",
  Deleting: "deleting",
  Running: "running"
};

// payloads/v2/user.ts
var UserPlanName = {
  Free: "free",
  Hobby: "hobby",
  Standard: "standard",
  Advanced: "advanced",
  Pro: "pro",
  Enterprise: (size) => `enterprise-${size}`
};
export {
  APINetworkDNSStatus,
  ApplicationId,
  ApplicationLanguage,
  ApplicationStatus,
  DeploymentState,
  FileType,
  UserId,
  UserPlanName
};
